source/common/stats/symbol_table_impl.h

```diff
@@ -400,6 +407,52 @@ class StatNameManagedStorage : public StatNameStorage {
   SymbolTable& symbol_table_;
 };
 
+/**
+ * Maintains storage for a collection of StatName objects. Like
+ * StatNameManagedStorage, this has an RAII usage model, taking
+ * care of decrementing ref-counts in the SymbolTable for all
+ * contained StatNames on destruction or on clear();
+ *
+ * Example usage:
+ *   StatNamePool pool(symbol_table);
+ *   StatName name1 = pool.add("name1");
+ *   StatName name2 = pool.add("name2");
+ *   uint8_t* storage = pool.addReturningStorage("name3");
+ *   StatName name3(storage);
+ */
+class StatNamePool {
+public:
+  explicit StatNamePool(SymbolTable& symbol_table) : symbol_table_(symbol_table) {}
+  ~StatNamePool() { clear(); }
+
+  /**
+   * Removes all StatNames from the pool.
+   */
+  void clear();
+
+  /**
+   * @param name the name to add the container.
+   * @return the StatName held in the container for this name.
+   */
+  StatName add(absl::string_view name);
+
+  /**
+   * Does essentially the same thing as add(), but returns the storage as a
+   * pointer which can later be used to create a StatName. This can be used
+   * to accumulate a vector of uint8_t* which can later be used to create
+   * StatName objects on demand.
+   *
+   * @param name the name to add the container.
+   * @return a pointer to the bytes held in the container for this name, suitable for
+   *         using to construct a StatName.
+   */
+  uint8_t* addReturningStorage(absl::string_view name);
+
+private:
+  SymbolTable& symbol_table_;
+  std::vector<StatNameStorage> storage_vector_;
```

Comment:
Why not just use `StatNameManagedStorage` here? Are the memory savings worth it?

Reply:
There can be a fair number of StatName objects added to a pool for HTTP Codes in #6733 again. IMO the cognitive cost of having a destructor is not that great in this case as this is in the stats impl. But if you prefer I can make this StatNameManagedStorage and remove the pool dtor, or just comment this tradeoff better.

Reply:
Sure it's fine, I would just add a comment on the tradeoff if you don't mind.

Reply:
done